# Introduction

Après avoir parcouru le fichier, nous allons commencer à le modifier

# Import des modules

In [1]:
import pandas as pd

# Lecture des données

In [2]:
df = pd.read_csv("datas.csv", index_col=0)

# Quelques rappels

### 10 première lignes

In [3]:
df.head(10)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_0  2006      5           8          1    705.0         700    917.0   
Vol_1  2006      5          10          3    705.0         700    920.0   
Vol_2  2006      5          12          5    700.0         700    922.0   
Vol_3  2006      5          14          7    700.0         700    910.0   
Vol_4  2006      5          16          2    724.0         700    921.0   
Vol_5  2006      5          18          4    658.0         700    909.0   
Vol_6  2006      5          20          6    657.0         700    908.0   
Vol_7  2006      5          22          1    653.0         700    904.0   
Vol_8  2006      5          24          3    658.0         700    914.0   
Vol_9  2006      5          26          5    800.0         700   1017.0   

       CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
Vol_0         915            US        560  ...      6        9          0   
Vol_1         915            US        560  ...      5       10          0   
Vol_2         915            US        560  ...      8       11          0   
Vol_3         915            US        560  ...      5        5          0   
Vol_4         915            US        560  ...      3        9          0   
Vol_5         915            US        560  ...      6       12          0   
Vol_6         915            US        560  ...      3       11          0   
Vol_7         915            US        560  ...      3        9          0   
Vol_8         915            US        560  ...      4        9          0   
Vol_9         915            US        560  ...      9       11          0   

       CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
Vol_0               NaN         0             0            0        0   
Vol_1               NaN         0             0            0        0   
Vol_2               NaN         0             0            0        0   
Vol_3               NaN         0             0            0        0   
Vol_4               NaN         0             0            0        0   
Vol_5               NaN         0             0            0        0   
Vol_6               NaN         0             0            0        0   
Vol_7               NaN         0             0            0        0   
Vol_8               NaN         0             0            0        0   
Vol_9               NaN         0            60            0        2   

       SecurityDelay  LateAircraftDelay  
Vol_0              0                  0  
Vol_1              0                  0  
Vol_2              0                  0  
Vol_3              0                  0  
Vol_4              0                  0  
Vol_5              0                  0  
Vol_6              0                  0  
Vol_7              0                  0  
Vol_8              0                  0  
Vol_9              0                  0  

[10 rows x 29 columns]

### 10 dernière lignes

In [4]:
df.tail(10)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_301450  2006      5          31          3   1922.0        1858   2151.0   
Vol_301451  2006      5          31          3   1501.0        1504   1841.0   
Vol_301452  2006      5          31          3   1851.0        1850   2024.0   
Vol_301453  2006      5          31          3    640.0         645   1408.0   
Vol_301454  2006      5          31          3   1151.0        1020   1833.0   
Vol_301455  2006      5          31          3   1838.0        1842   1943.0   
Vol_301456  2006      5          31          3    640.0         645   1307.0   
Vol_301457  2006      5          31          3   1222.0        1155   1422.0   
Vol_301458  2006      5          31          3   1830.0        1800   2057.0   
Vol_301459  2006      5          31          3   1615.0        1615   1755.0   

            CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  \
Vol_301450        2122            DL        371  ...      7       12   
Vol_301451        1826            DL        373  ...      7       40   
Vol_301452        2033            DL        376  ...      6       18   
Vol_301453        1400            DL        378  ...      8       41   
Vol_301454        1710            DL        379  ...     11       16   
Vol_301455        1945            DL        384  ...     12       16   
Vol_301456        1318            DL        385  ...      5       13   
Vol_301457        1345            DL        387  ...     11       24   
Vol_301458        2037            DL        389  ...      9       25   
Vol_301459        1804            DL        392  ...      6       14   

            Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay  \
Vol_301450          0               NaN         0            24            0   
Vol_301451          0               NaN         0             0            0   
Vol_301452          0               NaN         0             0            0   
Vol_301453          0               NaN         0             0            0   
Vol_301454          0               NaN         0            83            0   
Vol_301455          0               NaN         0             0            0   
Vol_301456          0               NaN         0             0            0   
Vol_301457          0               NaN         0            27            0   
Vol_301458          0               NaN         0             0            0   
Vol_301459          0               NaN         0             0            0   

           NASDelay  SecurityDelay  LateAircraftDelay  
Vol_301450        5              0                  0  
Vol_301451       15              0                  0  
Vol_301452        0              0                  0  
Vol_301453        0              0                  0  
Vol_301454        0              0                  0  
Vol_301455        0              0                  0  
Vol_301456        0              0                  0  
Vol_301457       10              0                  0  
Vol_301458       17              0                  3  
Vol_301459        0              0                  0  

[10 rows x 29 columns]

### Liste des colonnes

In [5]:
df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

### Dimensions du tableau

In [6]:
df.shape

(301460, 29)

### Première infos statistiques

In [7]:
df.describe()

Year     Month     DayofMonth      DayOfWeek        DepTime  \
count  301460.0  301460.0  301460.000000  301460.000000  297914.000000   
mean     2006.0       5.0      15.928073       3.743601    1338.087599   
std         0.0       0.0       8.946855       1.985304     473.382507   
min      2006.0       5.0       1.000000       1.000000       1.000000   
25%      2006.0       5.0       8.000000       2.000000     930.000000   
50%      2006.0       5.0      16.000000       4.000000    1327.000000   
75%      2006.0       5.0      24.000000       5.000000    1730.000000   
max      2006.0       5.0      31.000000       7.000000    2431.000000   

          CRSDepTime        ArrTime     CRSArrTime      FlightNum  \
count  301460.000000  297324.000000  301460.000000  301460.000000   
mean     1330.788725    1486.539930    1494.931002    2207.965889   
std       460.417894     496.297795     476.034379    2025.136577   
min        10.000000       1.000000       1.000000       1.000000   
25%       930.000000    1109.000000    1115.000000     576.000000   
50%      1323.000000    1516.000000    1520.000000    1476.000000   
75%      1720.000000    1909.000000    1905.000000    3640.000000   
max      2359.000000    2550.000000    2359.000000    9501.000000   

       ActualElapsedTime  ...       Distance         TaxiIn        TaxiOut  \
count      297324.000000  ...  301460.000000  301460.000000  301460.000000   
mean          125.134967  ...     730.104979       6.797406      15.426965   
std            71.050381  ...     577.823534      29.621836      10.887856   
min            16.000000  ...      24.000000       0.000000       0.000000   
25%            74.000000  ...     316.000000       4.000000      10.000000   
50%           106.000000  ...     576.000000       5.000000      13.000000   
75%           155.000000  ...     957.000000       7.000000      18.000000   
max          1511.000000  ...    4962.000000    1467.000000     292.000000   

           Cancelled       Diverted   CarrierDelay   WeatherDelay  \
count  301460.000000  301460.000000  301460.000000  301460.000000   
mean        0.011763       0.001957       2.879888       0.541531   
std         0.107817       0.044196      17.168730       7.330082   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000    1064.000000     646.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  301460.000000  301460.000000      301460.000000  
mean        3.194497       0.014264           4.011441  
std        14.445620       0.799616          18.106906  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         0.000000       0.000000           0.000000  
75%         0.000000       0.000000           0.000000  
max       495.000000     179.000000         515.000000  

[8 rows x 24 columns]

### Décompte des valeurs présentes

In [8]:
df.count()

Year                 301460
Month                301460
DayofMonth           301460
DayOfWeek            301460
DepTime              297914
CRSDepTime           301460
ArrTime              297324
CRSArrTime           301460
UniqueCarrier        301460
FlightNum            301460
TailNum              301460
ActualElapsedTime    297324
CRSElapsedTime       301460
AirTime              297324
ArrDelay             297324
DepDelay             297914
Origin               301460
Dest                 301460
Distance             301460
TaxiIn               301460
TaxiOut              301460
Cancelled            301460
CancellationCode       3546
Diverted             301460
CarrierDelay         301460
WeatherDelay         301460
NASDelay             301460
SecurityDelay        301460
LateAircraftDelay    301460
dtype: int64

# Gestion des manquants

### On peut supprimer les lignes avec une ou des données manquantes

In [9]:
df.dropna()

Empty DataFrame
Columns: [Year, Month, DayofMonth, DayOfWeek, DepTime, CRSDepTime, ArrTime, CRSArrTime, UniqueCarrier, FlightNum, TailNum, ActualElapsedTime, CRSElapsedTime, AirTime, ArrDelay, DepDelay, Origin, Dest, Distance, TaxiIn, TaxiOut, Cancelled, CancellationCode, Diverted, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay]
Index: []

[0 rows x 29 columns]

### On peut supprimer les lignes pour lesquelles CancellationCode est na 

In [10]:
df.dropna(subset=['CancellationCode'])

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_74      2006      5          12          5      NaN         615      NaN   
Vol_288     2006      5          12          5      NaN         630      NaN   
Vol_326     2006      5          30          2      NaN        1710      NaN   
Vol_483     2006      5           8          1      NaN        1245      NaN   
Vol_554     2006      5           9          2      NaN        1745      NaN   
...          ...    ...         ...        ...      ...         ...      ...   
Vol_301097  2006      5          30          2      NaN        2242      NaN   
Vol_301410  2006      5          30          2      NaN        2030      NaN   
Vol_301421  2006      5          30          2      NaN        1630      NaN   
Vol_301423  2006      5          30          2      NaN        1830      NaN   
Vol_301432  2006      5          31          3      NaN         900      NaN   

            CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  \
Vol_74             738            US        720  ...      0        0   
Vol_288            736            US       1159  ...      0        0   
Vol_326           1842            US       1189  ...      0        0   
Vol_483           1417            US       2035  ...      0        0   
Vol_554           1922            US       2045  ...      0        0   
...                ...           ...        ...  ...    ...      ...   
Vol_301097        2321            DL       1177  ...      0        0   
Vol_301410        2135            DL       1935  ...      0        0   
Vol_301421        1738            DL       1960  ...      0        0   
Vol_301423        1941            DL       1964  ...      0        0   
Vol_301432        1137            DL        271  ...      0        0   

            Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay  \
Vol_74              1                 B         0             0            0   
Vol_288             1                 B         0             0            0   
Vol_326             1                 A         0             0            0   
Vol_483             1                 A         0             0            0   
Vol_554             1                 C         0             0            0   
...               ...               ...       ...           ...          ...   
Vol_301097          1                 A         0             0            0   
Vol_301410          1                 A         0             0            0   
Vol_301421          1                 A         0             0            0   
Vol_301423          1                 C         0             0            0   
Vol_301432          1                 A         0             0            0   

           NASDelay  SecurityDelay  LateAircraftDelay  
Vol_74            0              0                  0  
Vol_288           0              0                  0  
Vol_326           0              0                  0  
Vol_483           0              0                  0  
Vol_554           0              0                  0  
...             ...            ...                ...  
Vol_301097        0              0                  0  
Vol_301410        0              0                  0  
Vol_301421        0              0                  0  
Vol_301423        0              0                  0  
Vol_301432        0              0                  0  

[3546 rows x 29 columns]

### On peut par exemple remplacer les valeurs manquante par un objet de notre choix

In [11]:
df.fillna(0)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_0       2006      5           8          1    705.0         700    917.0   
Vol_1       2006      5          10          3    705.0         700    920.0   
Vol_2       2006      5          12          5    700.0         700    922.0   
Vol_3       2006      5          14          7    700.0         700    910.0   
Vol_4       2006      5          16          2    724.0         700    921.0   
...          ...    ...         ...        ...      ...         ...      ...   
Vol_301455  2006      5          31          3   1838.0        1842   1943.0   
Vol_301456  2006      5          31          3    640.0         645   1307.0   
Vol_301457  2006      5          31          3   1222.0        1155   1422.0   
Vol_301458  2006      5          31          3   1830.0        1800   2057.0   
Vol_301459  2006      5          31          3   1615.0        1615   1755.0   

            CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  \
Vol_0              915            US        560  ...      6        9   
Vol_1              915            US        560  ...      5       10   
Vol_2              915            US        560  ...      8       11   
Vol_3              915            US        560  ...      5        5   
Vol_4              915            US        560  ...      3        9   
...                ...           ...        ...  ...    ...      ...   
Vol_301455        1945            DL        384  ...     12       16   
Vol_301456        1318            DL        385  ...      5       13   
Vol_301457        1345            DL        387  ...     11       24   
Vol_301458        2037            DL        389  ...      9       25   
Vol_301459        1804            DL        392  ...      6       14   

            Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay  \
Vol_0               0                 0         0             0            0   
Vol_1               0                 0         0             0            0   
Vol_2               0                 0         0             0            0   
Vol_3               0                 0         0             0            0   
Vol_4               0                 0         0             0            0   
...               ...               ...       ...           ...          ...   
Vol_301455          0                 0         0             0            0   
Vol_301456          0                 0         0             0            0   
Vol_301457          0                 0         0            27            0   
Vol_301458          0                 0         0             0            0   
Vol_301459          0                 0         0             0            0   

           NASDelay  SecurityDelay  LateAircraftDelay  
Vol_0             0              0                  0  
Vol_1             0              0                  0  
Vol_2             0              0                  0  
Vol_3             0              0                  0  
Vol_4             0              0                  0  
...             ...            ...                ...  
Vol_301455        0              0                  0  
Vol_301456        0              0                  0  
Vol_301457       10              0                  0  
Vol_301458       17              0                  3  
Vol_301459        0              0                  0  

[301460 rows x 29 columns]

In [12]:
df.fillna('toto')

Year  Month  DayofMonth  DayOfWeek DepTime  CRSDepTime ArrTime  \
Vol_0       2006      5           8          1     705         700     917   
Vol_1       2006      5          10          3     705         700     920   
Vol_2       2006      5          12          5     700         700     922   
Vol_3       2006      5          14          7     700         700     910   
Vol_4       2006      5          16          2     724         700     921   
...          ...    ...         ...        ...     ...         ...     ...   
Vol_301455  2006      5          31          3    1838        1842    1943   
Vol_301456  2006      5          31          3     640         645    1307   
Vol_301457  2006      5          31          3    1222        1155    1422   
Vol_301458  2006      5          31          3    1830        1800    2057   
Vol_301459  2006      5          31          3    1615        1615    1755   

            CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn TaxiOut  \
Vol_0              915            US        560  ...      6       9   
Vol_1              915            US        560  ...      5      10   
Vol_2              915            US        560  ...      8      11   
Vol_3              915            US        560  ...      5       5   
Vol_4              915            US        560  ...      3       9   
...                ...           ...        ...  ...    ...     ...   
Vol_301455        1945            DL        384  ...     12      16   
Vol_301456        1318            DL        385  ...      5      13   
Vol_301457        1345            DL        387  ...     11      24   
Vol_301458        2037            DL        389  ...      9      25   
Vol_301459        1804            DL        392  ...      6      14   

            Cancelled CancellationCode Diverted CarrierDelay WeatherDelay  \
Vol_0               0             toto        0            0            0   
Vol_1               0             toto        0            0            0   
Vol_2               0             toto        0            0            0   
Vol_3               0             toto        0            0            0   
Vol_4               0             toto        0            0            0   
...               ...              ...      ...          ...          ...   
Vol_301455          0             toto        0            0            0   
Vol_301456          0             toto        0            0            0   
Vol_301457          0             toto        0           27            0   
Vol_301458          0             toto        0            0            0   
Vol_301459          0             toto        0            0            0   

           NASDelay  SecurityDelay  LateAircraftDelay  
Vol_0             0              0                  0  
Vol_1             0              0                  0  
Vol_2             0              0                  0  
Vol_3             0              0                  0  
Vol_4             0              0                  0  
...             ...            ...                ...  
Vol_301455        0              0                  0  
Vol_301456        0              0                  0  
Vol_301457       10              0                  0  
Vol_301458       17              0                  3  
Vol_301459        0              0                  0  

[301460 rows x 29 columns]

### On peut remplacer les valeurs manquantes d'une colonne spécifique

In [13]:
df.CancellationCode.fillna('Z')

Vol_0         Z
Vol_1         Z
Vol_2         Z
Vol_3         Z
Vol_4         Z
             ..
Vol_301455    Z
Vol_301456    Z
Vol_301457    Z
Vol_301458    Z
Vol_301459    Z
Name: CancellationCode, Length: 301460, dtype: object

In [14]:
df.DepTime.fillna(df.DepTime.median())

Vol_0          705.0
Vol_1          705.0
Vol_2          700.0
Vol_3          700.0
Vol_4          724.0
               ...  
Vol_301455    1838.0
Vol_301456     640.0
Vol_301457    1222.0
Vol_301458    1830.0
Vol_301459    1615.0
Name: DepTime, Length: 301460, dtype: float64

### En l'état, on n'a pas modifié le DataFrame, on a juste affiché ce qu'est le DataFrame en remplaçant les valeurs manquantes.
### Pour réellement modifier le DataFrame, on utilise l'argument inplace=True

In [15]:
df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_0       2006      5           8          1    705.0         700    917.0   
Vol_1       2006      5          10          3    705.0         700    920.0   
Vol_2       2006      5          12          5    700.0         700    922.0   
Vol_3       2006      5          14          7    700.0         700    910.0   
Vol_4       2006      5          16          2    724.0         700    921.0   
...          ...    ...         ...        ...      ...         ...      ...   
Vol_301455  2006      5          31          3   1838.0        1842   1943.0   
Vol_301456  2006      5          31          3    640.0         645   1307.0   
Vol_301457  2006      5          31          3   1222.0        1155   1422.0   
Vol_301458  2006      5          31          3   1830.0        1800   2057.0   
Vol_301459  2006      5          31          3   1615.0        1615   1755.0   

            CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  \
Vol_0              915            US        560  ...      6        9   
Vol_1              915            US        560  ...      5       10   
Vol_2              915            US        560  ...      8       11   
Vol_3              915            US        560  ...      5        5   
Vol_4              915            US        560  ...      3        9   
...                ...           ...        ...  ...    ...      ...   
Vol_301455        1945            DL        384  ...     12       16   
Vol_301456        1318            DL        385  ...      5       13   
Vol_301457        1345            DL        387  ...     11       24   
Vol_301458        2037            DL        389  ...      9       25   
Vol_301459        1804            DL        392  ...      6       14   

            Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay  \
Vol_0               0               NaN         0             0            0   
Vol_1               0               NaN         0             0            0   
Vol_2               0               NaN         0             0            0   
Vol_3               0               NaN         0             0            0   
Vol_4               0               NaN         0             0            0   
...               ...               ...       ...           ...          ...   
Vol_301455          0               NaN         0             0            0   
Vol_301456          0               NaN         0             0            0   
Vol_301457          0               NaN         0            27            0   
Vol_301458          0               NaN         0             0            0   
Vol_301459          0               NaN         0             0            0   

           NASDelay  SecurityDelay  LateAircraftDelay  
Vol_0             0              0                  0  
Vol_1             0              0                  0  
Vol_2             0              0                  0  
Vol_3             0              0                  0  
Vol_4             0              0                  0  
...             ...            ...                ...  
Vol_301455        0              0                  0  
Vol_301456        0              0                  0  
Vol_301457       10              0                  0  
Vol_301458       17              0                  3  
Vol_301459        0              0                  0  

[301460 rows x 29 columns]

In [16]:
df.fillna(0, inplace=True)

In [17]:
df.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_0  2006      5           8          1    705.0         700    917.0   
Vol_1  2006      5          10          3    705.0         700    920.0   
Vol_2  2006      5          12          5    700.0         700    922.0   
Vol_3  2006      5          14          7    700.0         700    910.0   
Vol_4  2006      5          16          2    724.0         700    921.0   

       CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
Vol_0         915            US        560  ...      6        9          0   
Vol_1         915            US        560  ...      5       10          0   
Vol_2         915            US        560  ...      8       11          0   
Vol_3         915            US        560  ...      5        5          0   
Vol_4         915            US        560  ...      3        9          0   

       CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
Vol_0                 0         0             0            0        0   
Vol_1                 0         0             0            0        0   
Vol_2                 0         0             0            0        0   
Vol_3                 0         0             0            0        0   
Vol_4                 0         0             0            0        0   

       SecurityDelay  LateAircraftDelay  
Vol_0              0                  0  
Vol_1              0                  0  
Vol_2              0                  0  
Vol_3              0                  0  
Vol_4              0                  0  

[5 rows x 29 columns]

# Suppression de colonnes
Les colonnes Year et Month ne servent à rien puisque le fichier ne concerne que mai 2006

In [18]:
df.drop(['Year', 'Month'], axis=1, inplace=True)

In [19]:
df.shape

(301460, 27)

In [20]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1    705.0         700    917.0         915   
Vol_1          10          3    705.0         700    920.0         915   
Vol_2          12          5    700.0         700    922.0         915   
Vol_3          14          7    700.0         700    910.0         915   
Vol_4          16          2    724.0         700    921.0         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime  ...  TaxiIn  \
Vol_0            US        560  N405US              132.0  ...       6   
Vol_1            US        560  N784AU              135.0  ...       5   
Vol_2            US        560  N421US              142.0  ...       8   
Vol_3            US        560  N435US              130.0  ...       5   
Vol_4            US        560  N426US              117.0  ...       3   

       TaxiOut  Cancelled  CancellationCode Diverted CarrierDelay  \
Vol_0        9          0                 0        0            0   
Vol_1       10          0                 0        0            0   
Vol_2       11          0                 0        0            0   
Vol_3        5          0                 0        0            0   
Vol_4        9          0                 0        0            0   

       WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Vol_0             0         0              0                  0  
Vol_1             0         0              0                  0  
Vol_2             0         0              0                  0  
Vol_3             0         0              0                  0  
Vol_4             0         0              0                  0  

[5 rows x 27 columns]

### Exercice : supprimer les colonnes TaxiIn et TaxiOut

In [21]:
df.drop(['TaxiIn', 'TaxiOut'], axis=1, inplace=True)

In [22]:
df.shape

(301460, 25)

In [23]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1    705.0         700    917.0         915   
Vol_1          10          3    705.0         700    920.0         915   
Vol_2          12          5    700.0         700    922.0         915   
Vol_3          14          7    700.0         700    910.0         915   
Vol_4          16          2    724.0         700    921.0         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime  ...  Dest  \
Vol_0            US        560  N405US              132.0  ...   DCA   
Vol_1            US        560  N784AU              135.0  ...   DCA   
Vol_2            US        560  N421US              142.0  ...   DCA   
Vol_3            US        560  N435US              130.0  ...   DCA   
Vol_4            US        560  N426US              117.0  ...   DCA   

       Distance  Cancelled  CancellationCode Diverted CarrierDelay  \
Vol_0       857          0                 0        0            0   
Vol_1       857          0                 0        0            0   
Vol_2       857          0                 0        0            0   
Vol_3       857          0                 0        0            0   
Vol_4       857          0                 0        0            0   

       WeatherDelay  NASDelay SecurityDelay  LateAircraftDelay  
Vol_0             0         0             0                  0  
Vol_1             0         0             0                  0  
Vol_2             0         0             0                  0  
Vol_3             0         0             0                  0  
Vol_4             0         0             0                  0  

[5 rows x 25 columns]

# Suppression de lignes
Les vols annulés ne nous intéressent pas

In [24]:
df = df[df.Cancelled == 0]

In [25]:
df.shape

(297914, 25)

### Exercice : faire la même chose avec les vols déviés

In [26]:
df = df[df.Diverted == 0]

In [27]:
df.shape

(297324, 25)

### Les colonnes Diverted, Cancelled et CancellationCode sont devenues inutiles du fait de l'extraction précédente

In [28]:
df.drop(['Diverted', 'Cancelled', 'CancellationCode'], axis=1, inplace=True)

In [29]:
df.shape

(297324, 22)

In [30]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1    705.0         700    917.0         915   
Vol_1          10          3    705.0         700    920.0         915   
Vol_2          12          5    700.0         700    922.0         915   
Vol_3          14          7    700.0         700    910.0         915   
Vol_4          16          2    724.0         700    921.0         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime  ...  ArrDelay  \
Vol_0            US        560  N405US              132.0  ...       2.0   
Vol_1            US        560  N784AU              135.0  ...       5.0   
Vol_2            US        560  N421US              142.0  ...       7.0   
Vol_3            US        560  N435US              130.0  ...      -5.0   
Vol_4            US        560  N426US              117.0  ...       6.0   

       DepDelay  Origin  Dest Distance CarrierDelay  WeatherDelay  NASDelay  \
Vol_0       5.0     PBI   DCA      857            0             0         0   
Vol_1       5.0     PBI   DCA      857            0             0         0   
Vol_2       0.0     PBI   DCA      857            0             0         0   
Vol_3       0.0     PBI   DCA      857            0             0         0   
Vol_4      24.0     PBI   DCA      857            0             0         0   

       SecurityDelay  LateAircraftDelay  
Vol_0              0                  0  
Vol_1              0                  0  
Vol_2              0                  0  
Vol_3              0                  0  
Vol_4              0                  0  

[5 rows x 22 columns]

### Pour finir, d'autres colonnes ne nous intéresseront pas : FlightNum, TailNum

In [31]:
df.drop(['FlightNum', 'TailNum'], axis=1, inplace=True)

In [32]:
df.shape

(297324, 20)

In [33]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1    705.0         700    917.0         915   
Vol_1          10          3    705.0         700    920.0         915   
Vol_2          12          5    700.0         700    922.0         915   
Vol_3          14          7    700.0         700    910.0         915   
Vol_4          16          2    724.0         700    921.0         915   

      UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
Vol_0            US              132.0           135.0    117.0       2.0   
Vol_1            US              135.0           135.0    120.0       5.0   
Vol_2            US              142.0           135.0    123.0       7.0   
Vol_3            US              130.0           135.0    120.0      -5.0   
Vol_4            US              117.0           135.0    105.0       6.0   

       DepDelay Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
Vol_0       5.0    PBI  DCA       857             0             0         0   
Vol_1       5.0    PBI  DCA       857             0             0         0   
Vol_2       0.0    PBI  DCA       857             0             0         0   
Vol_3       0.0    PBI  DCA       857             0             0         0   
Vol_4      24.0    PBI  DCA       857             0             0         0   

       SecurityDelay  LateAircraftDelay  
Vol_0              0                  0  
Vol_1              0                  0  
Vol_2              0                  0  
Vol_3              0                  0  
Vol_4              0                  0

# Tri des lignes

### On souhaite trier les lignes selon l'heure de départ

In [34]:
df.sort_values(by='DepTime', inplace=True, ascending=True)

In [35]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_3373            25          4      1.0        2355    741.0         723   
Vol_64530            5          5      1.0        2351    200.0         146   
Vol_256163           4          4      1.0        2355     56.0          49   
Vol_117443          14          7      1.0        1920    257.0        2249   
Vol_140794          30          2      1.0        2105    158.0        2312   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_3373              US              280.0           268.0    251.0   
Vol_64530             YV              119.0           115.0     91.0   
Vol_256163            AS               55.0            54.0     42.0   
Vol_117443            XE              116.0           149.0    100.0   
Vol_140794            UA               57.0            67.0     38.0   

            ArrDelay  DepDelay Origin Dest  Distance  CarrierDelay  \
Vol_3373        18.0       6.0    LAS  MIA      2175             6   
Vol_64530       14.0      10.0    LAS  MFR       599             0   
Vol_256163       7.0       6.0    ANC  FAI       261             0   
Vol_117443     248.0     281.0    IAH  GSP       838            69   
Vol_140794     166.0     176.0    ORD  CVG       264             0   

            WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Vol_3373               0        12              0                  0  
Vol_64530              0         0              0                  0  
Vol_256163             0         0              0                  0  
Vol_117443             0         0              0                179  
Vol_140794            15         0              0                151

### Exercice : trouver comment trier à la fois par jour du mois et heure de départ

In [36]:
df.sort_values(by=['DayofMonth', 'DepTime'], inplace=True, ascending=True)

In [37]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_8424             1          1      1.0        2359    730.0         733   
Vol_238734           1          1      4.0        2237    110.0        2340   
Vol_262396           1          1      8.0        2355    819.0         800   
Vol_238703           1          1      8.0        2355    530.0         530   
Vol_253414           1          1      8.0        2215    129.0        2335   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_8424              US              269.0           274.0    248.0   
Vol_238734            AA               66.0            63.0     53.0   
Vol_262396            B6              311.0           305.0    466.0   
Vol_238703            AA              202.0           215.0    190.0   
Vol_253414            AA               81.0            80.0     63.0   

            ArrDelay  DepDelay Origin Dest  Distance  CarrierDelay  \
Vol_8424        -3.0       2.0    LAS  FLL      2174             0   
Vol_238734      90.0      87.0    ORD  STL       258             0   
Vol_262396      19.0      13.0    PDX  JFK      2454            13   
Vol_238703       0.0      13.0    SEA  DFW      1660             0   
Vol_253414     114.0     113.0    ORD  MCI       403             0   

            WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Vol_8424               0         0              0                  0  
Vol_238734             0         3              0                 87  
Vol_262396             0         6              0                  0  
Vol_238703             0         0              0                  0  
Vol_253414             0         3              0                111

In [38]:
df.DepTime.diff()

Vol_8424       NaN
Vol_238734     3.0
Vol_262396     4.0
Vol_238703     0.0
Vol_253414     0.0
              ... 
Vol_135902     0.0
Vol_16401      1.0
Vol_16425      0.0
Vol_150941     1.0
Vol_211153    42.0
Name: DepTime, Length: 297324, dtype: float64

In [39]:
df.DepTime.pct_change()

Vol_8424           NaN
Vol_238734    3.000000
Vol_262396    1.000000
Vol_238703    0.000000
Vol_253414    0.000000
                ...   
Vol_135902    0.000000
Vol_16401     0.000424
Vol_16425     0.000000
Vol_150941    0.000424
Vol_211153    0.017812
Name: DepTime, Length: 297324, dtype: float64

# Création de colonnes
On souhaite créer une colonne renseignant, à partir des différents types de retard, le retard total de chaque vol

### Création de la colonne du retard total (il existe plusieurs méthodes)

In [40]:
df['TotalDelay'] = df.CarrierDelay + df.WeatherDelay + df.NASDelay + df.SecurityDelay + df.LateAircraftDelay

In [41]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_8424             1          1      1.0        2359    730.0         733   
Vol_238734           1          1      4.0        2237    110.0        2340   
Vol_262396           1          1      8.0        2355    819.0         800   
Vol_238703           1          1      8.0        2355    530.0         530   
Vol_253414           1          1      8.0        2215    129.0        2335   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  ...  \
Vol_8424              US              269.0           274.0    248.0  ...   
Vol_238734            AA               66.0            63.0     53.0  ...   
Vol_262396            B6              311.0           305.0    466.0  ...   
Vol_238703            AA              202.0           215.0    190.0  ...   
Vol_253414            AA               81.0            80.0     63.0  ...   

            DepDelay  Origin Dest Distance  CarrierDelay  WeatherDelay  \
Vol_8424         2.0     LAS  FLL     2174             0             0   
Vol_238734      87.0     ORD  STL      258             0             0   
Vol_262396      13.0     PDX  JFK     2454            13             0   
Vol_238703      13.0     SEA  DFW     1660             0             0   
Vol_253414     113.0     ORD  MCI      403             0             0   

            NASDelay  SecurityDelay  LateAircraftDelay  TotalDelay  
Vol_8424           0              0                  0           0  
Vol_238734         3              0                 87          90  
Vol_262396         6              0                  0          19  
Vol_238703         0              0                  0           0  
Vol_253414         3              0                111         114  

[5 rows x 21 columns]

### Exercice : Chercher comment créer la colonne caractérisant si l'avion est en retard ou non (il existe plusieurs méthodes)

In [42]:
df['Delayed'] = df.TotalDelay > 0

In [43]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_8424             1          1      1.0        2359    730.0         733   
Vol_238734           1          1      4.0        2237    110.0        2340   
Vol_262396           1          1      8.0        2355    819.0         800   
Vol_238703           1          1      8.0        2355    530.0         530   
Vol_253414           1          1      8.0        2215    129.0        2335   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  ...  \
Vol_8424              US              269.0           274.0    248.0  ...   
Vol_238734            AA               66.0            63.0     53.0  ...   
Vol_262396            B6              311.0           305.0    466.0  ...   
Vol_238703            AA              202.0           215.0    190.0  ...   
Vol_253414            AA               81.0            80.0     63.0  ...   

            Origin  Dest Distance CarrierDelay  WeatherDelay  NASDelay  \
Vol_8424       LAS   FLL     2174            0             0         0   
Vol_238734     ORD   STL      258            0             0         3   
Vol_262396     PDX   JFK     2454           13             0         6   
Vol_238703     SEA   DFW     1660            0             0         0   
Vol_253414     ORD   MCI      403            0             0         3   

            SecurityDelay  LateAircraftDelay  TotalDelay  Delayed  
Vol_8424                0                  0           0    False  
Vol_238734              0                 87          90     True  
Vol_262396              0                  0          19     True  
Vol_238703              0                  0           0    False  
Vol_253414              0                111         114     True  

[5 rows x 22 columns]

In [44]:
df.Delayed.memory_usage() / 1024 / 1024

12.551952362060547

# Supprimer les différentes colonnes détaillant le retard
On ne s'intéresse qu'au retard total, on peut donc supprimer les colonnes détaillant les raisons du retard

In [45]:
df.drop(['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'], axis=1, inplace=True)

In [46]:
df.shape

(297324, 17)

In [47]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_8424             1          1      1.0        2359    730.0         733   
Vol_238734           1          1      4.0        2237    110.0        2340   
Vol_262396           1          1      8.0        2355    819.0         800   
Vol_238703           1          1      8.0        2355    530.0         530   
Vol_253414           1          1      8.0        2215    129.0        2335   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_8424              US              269.0           274.0    248.0   
Vol_238734            AA               66.0            63.0     53.0   
Vol_262396            B6              311.0           305.0    466.0   
Vol_238703            AA              202.0           215.0    190.0   
Vol_253414            AA               81.0            80.0     63.0   

            ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_8424        -3.0       2.0    LAS  FLL      2174           0    False  
Vol_238734      90.0      87.0    ORD  STL       258          90     True  
Vol_262396      19.0      13.0    PDX  JFK      2454          19     True  
Vol_238703       0.0      13.0    SEA  DFW      1660           0    False  
Vol_253414     114.0     113.0    ORD  MCI       403         114     True

# Modifier une colonne
Les colonnes DepTime, CRSDepTime, ArrTime et CRSArrTime sont formatées de la façon suivant : hhmm. Par exemple, un DepTime de 724 correspond à 7h24 le matin. On souhaite modifier cette colonne pour avoir le temps en minutes. Par exemple, 7h24 devient 444.

### On commence par modifier la colonne DepTime

In [48]:
df['DepTime'] = (df.DepTime // 100) * 60 + df.DepTime % 100

In [49]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_135902          31          3   1436.0        2320    539.0         502   
Vol_16401           31          3   1437.0        2358    746.0         753   
Vol_16425           31          3   1437.0        2359    106.0         107   
Vol_150941          31          3   1438.0        2345    728.0         725   
Vol_211153          31          3   1440.0        2229     56.0        2329   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_135902            UA              223.0           222.0    200.0   
Vol_16401             US              289.0           295.0    261.0   
Vol_16425             US               69.0            68.0     42.0   
Vol_150941            DL              270.0           280.0    248.0   
Vol_211153            NW               56.0            60.0     40.0   

            ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_135902      37.0      36.0    PDX  ORD      1739          37     True  
Vol_16401       -7.0      -1.0    LAS  EWR      2227           0    False  
Vol_16425       -1.0      -2.0    LAS  SAN       258           0    False  
Vol_150941       3.0      13.0    SEA  ATL      2182           0    False  
Vol_211153      87.0      91.0    MSP  FSD       197          87     True

### Exercice : Faire la même transformation pour CRSDepTime, ArrTime et CRSArrTime

In [50]:
df['CRSDepTime'] = (df.CRSDepTime // 100) * 60 + df.CRSDepTime % 100

In [51]:
df['ArrTime'] = (df.ArrTime // 100) * 60 + df.ArrTime % 100

In [52]:
df['CRSArrTime'] = (df.CRSArrTime // 100) * 60 + df.CRSArrTime % 100

### Cas du AirTime

In [53]:
df.describe()

DayofMonth      DayOfWeek        DepTime     CRSDepTime  \
count  297324.000000  297324.000000  297324.000000  297324.000000   
mean       15.903449       3.746162     814.659257     808.527169   
std         8.943123       1.987111     283.953208     276.951876   
min         1.000000       1.000000       1.000000      10.000000   
25%         8.000000       2.000000     570.000000     570.000000   
50%        16.000000       4.000000     807.000000     800.000000   
75%        24.000000       5.000000    1050.000000    1038.000000   
max        31.000000       7.000000    1471.000000    1439.000000   

             ArrTime     CRSArrTime  ActualElapsedTime  CRSElapsedTime  \
count  297324.000000  297324.000000      297324.000000   297324.000000   
mean      903.665355     908.016790         125.134967      126.755432   
std       297.849917     285.798374          71.050381       70.514544   
min         1.000000       1.000000          16.000000       19.000000   
25%       669.000000     675.000000          74.000000       75.000000   
50%       916.000000     919.000000         106.000000      108.000000   
75%      1149.000000    1144.000000         155.000000      156.000000   
max      1550.000000    1439.000000        1511.000000      640.000000   

             AirTime       ArrDelay       DepDelay       Distance  \
count  297324.000000  297324.000000  297324.000000  297324.000000   
mean      102.644391       6.928381       8.548846     732.252496   
std        75.888877      33.664616      30.402858     578.786571   
min     -1416.000000    -240.000000   -1020.000000      24.000000   
25%        54.000000      -9.000000      -4.000000     317.000000   
50%        84.000000      -2.000000       0.000000     576.000000   
75%       131.000000      10.000000       8.000000     958.000000   
max      1484.000000    1078.000000    1070.000000    4962.000000   

          TotalDelay  
count  297324.000000  
mean       10.789654  
std        31.205019  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max      1078.000000

### A votre avis, que se passe-t-il et comment corriger cela ?

In [54]:
df[df.AirTime < 0]

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_152810           1          1   1355.0        1360   1475.0          49   
Vol_159493           1          1   1356.0        1360   1470.0          41   
Vol_155887           1          1   1380.0        1355   1448.0        1420   
Vol_154949           1          1   1470.0        1380   1520.0           1   
Vol_160527           2          2   1365.0        1365   1460.0          18   
...                ...        ...      ...         ...      ...         ...   
Vol_151344          29          1   1471.0        1380   1517.0          10   
Vol_160830          30          2   1378.0        1375   1449.0          12   
Vol_152825          31          3   1360.0        1360   1482.0          49   
Vol_157389          31          3   1376.0        1370   1453.0           7   
Vol_153939          31          3   1390.0        1266   1456.0        1361   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_152810            EV              120.0           129.0  -1337.0   
Vol_159493            EV              114.0           121.0  -1345.0   
Vol_155887            EV              128.0           125.0  -1338.0   
Vol_154949            EV               50.0            61.0  -1409.0   
Vol_160527            EV               95.0            93.0  -1374.0   
...                  ...                ...             ...      ...   
Vol_151344            EV               46.0            70.0  -1405.0   
Vol_160830            EV              131.0           137.0  -1336.0   
Vol_152825            EV              122.0           129.0  -1332.0   
Vol_157389            EV               77.0            77.0  -1386.0   
Vol_153939            EV               66.0            95.0  -1383.0   

            ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_152810     -14.0      -5.0    ATL  SYR       793           0    False  
Vol_159493     -11.0      -4.0    ATL  ROC       749           0    False  
Vol_155887      28.0      25.0    ATL  HOU       696          28     True  
Vol_154949      79.0      90.0    ATL  AGS       143          79     True  
Vol_160527       2.0       0.0    ATL  PHF       508           0    False  
...              ...       ...    ...  ...       ...         ...      ...  
Vol_151344      67.0      91.0    ATL  AVL       164          67     True  
Vol_160830      -3.0       3.0    ATL  OMA       821           0    False  
Vol_152825      -7.0       0.0    ATL  SYR       793           0    False  
Vol_157389       6.0       6.0    ATL  ILM       377           0    False  
Vol_153939      95.0     124.0    SLC  MSO       436          95     True  

[125 rows x 17 columns]

In [55]:
df.AirTime = [x + 1440 if x < 0 else x for x in df.AirTime]

In [56]:
df.loc['Vol_152810']

DayofMonth               1
DayOfWeek                1
DepTime               1355
CRSDepTime            1360
ArrTime               1475
CRSArrTime              49
UniqueCarrier           EV
ActualElapsedTime      120
CRSElapsedTime         129
AirTime                103
ArrDelay               -14
DepDelay                -5
Origin                 ATL
Dest                   SYR
Distance               793
TotalDelay               0
Delayed              False
Name: Vol_152810, dtype: object

In [57]:
df[df.AirTime > 18*60]

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_264988          15          1    340.0        1360    411.0        1430   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_264988            B6             1511.0            70.0   1484.0   

            ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_264988     421.0   -1020.0    JFK  BTV       267         421     True

In [58]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_8424             1          1      1.0        1439    450.0         453   
Vol_238734           1          1      4.0        1357     70.0        1420   
Vol_262396           1          1      8.0        1435    499.0         480   
Vol_238703           1          1      8.0        1435    330.0         330   
Vol_253414           1          1      8.0        1335     89.0        1415   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_8424              US              269.0           274.0    248.0   
Vol_238734            AA               66.0            63.0     53.0   
Vol_262396            B6              311.0           305.0    466.0   
Vol_238703            AA              202.0           215.0    190.0   
Vol_253414            AA               81.0            80.0     63.0   

            ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_8424        -3.0       2.0    LAS  FLL      2174           0    False  
Vol_238734      90.0      87.0    ORD  STL       258          90     True  
Vol_262396      19.0      13.0    PDX  JFK      2454          19     True  
Vol_238703       0.0      13.0    SEA  DFW      1660           0    False  
Vol_253414     114.0     113.0    ORD  MCI       403         114     True

# Synthèse des retards par compagnie
La méthode groupby va nous permettre d'extraire facilement des informations par compagnie : https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html

### Création d'un nouveau DataFrame

In [84]:
df_comp = pd.DataFrame(index=df.UniqueCarrier.unique())

In [85]:
df.UniqueCarrier.unique()

array(['US', 'AA', 'B6', 'NW', 'CO', 'YV', 'DL', 'MQ', 'AS', 'UA', 'XE',
       'FL', 'OO', 'HA', 'AQ', 'EV', 'OH', 'WN', 'F9', 'TZ'], dtype=object)

In [86]:
df_comp

Empty DataFrame
Columns: []
Index: [US, AA, B6, NW, CO, YV, DL, MQ, AS, UA, XE, FL, OO, HA, AQ, EV, OH, WN, F9, TZ]

In [87]:
df_comp.sort_index(inplace=True)

In [88]:
df_comp

Empty DataFrame
Columns: []
Index: [AA, AQ, AS, B6, CO, DL, EV, F9, FL, HA, MQ, NW, OH, OO, TZ, UA, US, WN, XE, YV]

### Première colonne du DataFrame

In [89]:
df_comp['compte'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').count()

In [90]:
df_comp.head()

compte
AA   27349
AQ    1800
AS    6688
B6    5985
CO   12904

### Exercice : Continuer à remplir le DataFrame avec la moyenne, l'écart-type, le minimum, le quantile 25%, le quantile 50%, le quantile 75% et le maximum

In [91]:
df_comp['moyenne'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').mean()

In [92]:
df_comp['ecarttype'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').std()

In [93]:
df_comp['minimum'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').min()

In [94]:
df_comp['Q25'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').quantile(0.25)

In [95]:
df_comp['mediane'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').median()

In [96]:
df_comp['Q75'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').quantile(0.75)

In [97]:
df_comp['maximum'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').max()

In [98]:
df_comp

compte    moyenne  ecarttype  minimum  Q25  mediane   Q75  maximum
AA   27349  13.237486  36.670329        0  0.0        0   0.0      993
AQ    1800   3.396111  13.197027        0  0.0        0   0.0      235
AS    6688   7.711872  26.100351        0  0.0        0   0.0      523
B6    5985   8.839599  28.645614        0  0.0        0   0.0      653
CO   12904  13.964585  35.367171        0  0.0        0  15.0      592
DL   20666   7.618891  23.805501        0  0.0        0   0.0      560
EV   11776  12.336872  32.868313        0  0.0        0   0.0     1061
F9    3842   5.354243  18.497668        0  0.0        0   0.0      471
FL    9894  10.901759  29.223487        0  0.0        0   0.0      561
HA    2060   1.928155  12.563377        0  0.0        0   0.0      283
MQ   22864  16.045224  36.974729        0  0.0        0  19.0      825
NW   18176   8.419509  34.723612        0  0.0        0   0.0     1078
OH   11516   9.135637  27.604238        0  0.0        0   0.0      535
OO   22776   8.804970  27.871130        0  0.0        0   0.0      462
TZ     765  18.994771  41.483328        0  0.0        0  22.0      293
UA   20960  15.676002  40.588612        0  0.0        0   0.0      526
US   21122   8.033283  23.675080        0  0.0        0   0.0      622
WN   45815   8.076023  22.976132        0  0.0        0   0.0      410
XE   18327  14.578545  36.235645        0  0.0        0  15.0      489
YV   12039  11.999252  33.567506        0  0.0        0   0.0      467

# Opérations plus complexes dans un groupby

### Application de la fonction finale à plusieurs colonnes

In [99]:
df.loc[:, ['UniqueCarrier', 'TotalDelay', 'CRSElapsedTime']].groupby('UniqueCarrier').max()

TotalDelay  CRSElapsedTime
UniqueCarrier                            
AA                    993           549.0
AQ                    235           354.0
AS                    523           414.0
B6                    653           400.0
CO                    592           640.0
DL                    560           579.0
EV                   1061           226.0
F9                    471           325.0
FL                    561           335.0
HA                    283           385.0
MQ                    825           209.0
NW                   1078           495.0
OH                    535           196.0
OO                    462           236.0
TZ                    293           392.0
UA                    526           561.0
US                    622           425.0
WN                    410           375.0
XE                    489           239.0
YV                    467           232.0

### grouper selon plusieurs colonnes

In [100]:
df.loc[:, ['Origin', 'Dest', 'TotalDelay']].groupby(['Origin', 'Dest']).max()

TotalDelay
Origin Dest            
ABE    ATL          394
       CLE           60
       CLT            0
       CVG            0
       JFK            0
...                 ...
YAK    CDV           22
       JNU            0
YUM    IPL           55
       LAX          100
       PHX          219

[4003 rows x 1 columns]

### Utilisation d'une lambda fonction

In [ ]:
df_AA = df[df.UniqueCarrier == 'AA']

In [ ]:
df_AQ = df[df.UniqueCarrier == 'AQ']

In [104]:
df.loc[:, ['UniqueCarrier', 'TotalDelay', 'CRSElapsedTime']].groupby('UniqueCarrier').apply(lambda x: 100*(x.TotalDelay / x.CRSElapsedTime).quantile(0.75))

UniqueCarrier
AA     0.000000
AQ     0.000000
AS     0.000000
B6     0.000000
CO     5.573770
DL     0.000000
EV     0.000000
F9     0.000000
FL     0.000000
HA     0.000000
MQ    21.186441
NW     0.000000
OH     0.000000
OO     0.000000
TZ    12.142857
UA     0.000000
US     0.000000
WN     0.000000
XE     8.937970
YV     0.000000
dtype: float64

# Merge
On souhaite ajouter les informations df df_comp dans notre DataFrame de départ

In [105]:
df_full = pd.merge(df, df_comp, how='left', left_on='UniqueCarrier', right_index=True)

In [106]:
df_full

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_8424             1          1      1.0        1439    450.0         453   
Vol_238734           1          1      4.0        1357     70.0        1420   
Vol_262396           1          1      8.0        1435    499.0         480   
Vol_238703           1          1      8.0        1435    330.0         330   
Vol_253414           1          1      8.0        1335     89.0        1415   
...                ...        ...      ...         ...      ...         ...   
Vol_135902          31          3   1436.0        1400    339.0         302   
Vol_16401           31          3   1437.0        1438    466.0         473   
Vol_16425           31          3   1437.0        1439     66.0          67   
Vol_150941          31          3   1438.0        1425    448.0         445   
Vol_211153          31          3   1440.0        1349     56.0        1409   

           UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  ...  \
Vol_8424              US              269.0           274.0    248.0  ...   
Vol_238734            AA               66.0            63.0     53.0  ...   
Vol_262396            B6              311.0           305.0    466.0  ...   
Vol_238703            AA              202.0           215.0    190.0  ...   
Vol_253414            AA               81.0            80.0     63.0  ...   
...                  ...                ...             ...      ...  ...   
Vol_135902            UA              223.0           222.0    200.0  ...   
Vol_16401             US              289.0           295.0    261.0  ...   
Vol_16425             US               69.0            68.0     42.0  ...   
Vol_150941            DL              270.0           280.0    248.0  ...   
Vol_211153            NW               56.0            60.0     40.0  ...   

            TotalDelay  Delayed compte    moyenne  ecarttype  minimum  Q25  \
Vol_8424             0    False  21122   8.033283  23.675080        0  0.0   
Vol_238734          90     True  27349  13.237486  36.670329        0  0.0   
Vol_262396          19     True   5985   8.839599  28.645614        0  0.0   
Vol_238703           0    False  27349  13.237486  36.670329        0  0.0   
Vol_253414         114     True  27349  13.237486  36.670329        0  0.0   
...                ...      ...    ...        ...        ...      ...  ...   
Vol_135902          37     True  20960  15.676002  40.588612        0  0.0   
Vol_16401            0    False  21122   8.033283  23.675080        0  0.0   
Vol_16425            0    False  21122   8.033283  23.675080        0  0.0   
Vol_150941           0    False  20666   7.618891  23.805501        0  0.0   
Vol_211153          87     True  18176   8.419509  34.723612        0  0.0   

            mediane  Q75  maximum  
Vol_8424          0  0.0      622  
Vol_238734        0  0.0      993  
Vol_262396        0  0.0      653  
Vol_238703        0  0.0      993  
Vol_253414        0  0.0      993  
...             ...  ...      ...  
Vol_135902        0  0.0      526  
Vol_16401         0  0.0      622  
Vol_16425         0  0.0      622  
Vol_150941        0  0.0      560  
Vol_211153        0  0.0     1078  

[297324 rows x 25 columns]

# Bonus : utilisation avec scikit-learn

In [107]:
from sklearn.preprocessing import MinMaxScaler

In [108]:
scaler = MinMaxScaler()

In [109]:
scaler.fit_transform(df[['DepTime', 'ArrTime']])

array([[0.        , 0.28986443],
       [0.00204082, 0.04454487],
       [0.0047619 , 0.32149774],
       ...,
       [0.97687075, 0.04196256],
       [0.97755102, 0.28857327],
       [0.97891156, 0.03550678]])

### Exercice : appliquez StandardScaler plutôt que MinMaxScaler

In [110]:
from sklearn.preprocessing import StandardScaler

In [111]:
scaler = StandardScaler()

In [112]:
scaler.fit_transform(df[['DepTime', 'ArrTime']])

array([[-2.86547432, -1.52313663],
       [-2.85490918, -2.7989491 ],
       [-2.84082233, -1.35862397],
       ...,
       [ 2.19170544, -2.81237871],
       [ 2.19522715, -1.52985143],
       [ 2.20227058, -2.84595272]])

### train_test_split

In [113]:
from sklearn.model_selection import train_test_split

In [114]:
X_train, X_test, y_train, y_test = train_test_split(df.drop(['Delayed', 'TotalDelay'], axis=1), df.TotalDelay)

In [115]:
X_train.shape

(222993, 15)

In [116]:
X_test.shape

(74331, 15)

In [117]:
y_train.shape

(222993,)

In [118]:
y_test.shape

(74331,)

# Sauvegarde des données
la méthode to_csv permet d'écrire le DataFrame modifié au format csv. L'argument index=True précise que l'on souhaite écrire l'index dans le fichier

In [120]:
df.to_csv('datas_cleaned.csv', index=True)